addons/controllers/csi/csiconfig_controller.go

```diff
@@ -0,0 +1,196 @@
+	}
+
+	return r.reconcileCSIConfig(ctx, csiConfig)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *CSIConfigReconciler) SetupWithManager(_ context.Context, mgr ctrl.Manager,
+	options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&csiv1alpha1.CSIConfig{}).
+		WithOptions(options).
+		Complete(r)
+}
+
+func (r *CSIConfigReconciler) reconcileCSIConfig(ctx context.Context,
+	csiCfg *csiv1alpha1.CSIConfig) (result ctrl.Result, retErr error) {
+
+	logger := log.FromContext(ctx)
+
+	patchHelper, err := clusterapipatchutil.NewHelper(csiCfg, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if retErr != nil {
+			// don't modify CSIConfig if there is an error
+			return
+		}
+
+		if err := patchHelper.Patch(ctx, csiCfg); err != nil {
+			logger.Error(err, "Failed to patch CSIConfig")
+			return
+		}
+	}()
+
+	if !csiCfg.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, nil // deleted
+	}
+
+	if result, err = r.reconcileCSIConfigNormal(ctx, csiCfg); err != nil {
+		logger.Error(err, "Error reconciling CSIConfig")
+		return result, err
+	}
+
+	return result, nil
+}
+
+func (r *CSIConfigReconciler) reconcileCSIConfigNormal(ctx context.Context,
+	csiCfg *csiv1alpha1.CSIConfig) (result ctrl.Result, retErr error) {
+
+	logger := log.FromContext(ctx)
+
+	// add owner reference to CSIConfig if not already added
+	cluster, err := r.getOwnerCluster(ctx, csiCfg)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	ownerRef := metav1.OwnerReference{
+		APIVersion: clusterapiv1beta1.GroupVersion.String(),
+		Kind:       cluster.Kind,
+		Name:       cluster.Name,
+		UID:        cluster.UID,
+	}
+
+	if !clusterapiutil.HasOwnerRef(csiCfg.OwnerReferences, ownerRef) {
+		csiCfg.OwnerReferences = clusterapiutil.EnsureOwnerRef(csiCfg.OwnerReferences, ownerRef)
+	}
+
+	addonName := ""
+	switch csiCfg.Spec.VSphereCSI.Mode {
+	case CSINonParavirtualMode:
+		addonName = "vsphere-csi"
+	case CSIParavirtualMode:
+		addonName = "vsphere-pv-csi"
+	}
+
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.GenerateDataValueSecretName(cluster.Name, addonName),
+			Namespace: csiCfg.Namespace},
+		Type: v1.SecretTypeOpaque,
+	}
+
+	secret.SetOwnerReferences([]metav1.OwnerReference{ownerRef})
+
+	mutateFn := func() error {
+		secret.Data = make(map[string][]byte)
+		dvs, err := r.mapCSIConfigToDataValues(ctx, csiCfg)
+		if err != nil {
+			logger.Error(err, "Error while mapping CSIConfig to data values")
+			return err
+		}
+		yamlBytes, err := yaml.Marshal(dvs)
+		if err != nil {
+			logger.Error(err, "Error marshaling CSI config data values to yaml")
+			return err
+		}
+		secret.Data[constants.TKGDataValueFileName] = yamlBytes
+		return nil
+	}
+
+	opResult, err := controllerutil.CreateOrPatch(ctx, r.Client, secret, mutateFn)
+
+	if err != nil {
+		logger.Error(err, "Error creating or patching CSIConfig data values secret")
+		return result, err
```

Comment:
What is the value of ```result```? I don't see it being set anywhere. Is that always ```ctrl.Result{}```?

Reply:
updated